### PR TITLE
Add idempotency middleware

### LINKED ISSUE #185: Add nonce/idempotency key to requests
This will require:
- instructions for clients on how to set the key per request
- decide if it's a header, query param or part of the request body
- handle the key in the backend and reject if a request with the same key is received

### PR DESCRIPTION
**Pull request by nanuuki**
This adds basic idempotency middleware for `POST` endpoints:
- only the idempotency key is stored, so the payload is not examined when checking the received idempotency key and hence only `409` can be returned (key already used). A `422` response would require storing the payload.
- there are 3 supported "databases" (see README for configuration):
  - `local`: in-memory, single instance only, mainly for testing purposes
  - `shared`:  use the same db as main app
  - `redis`: use a separate Redis instance
- keys expire after 1 hour, meaning the same key can be used after >1 hour after previous usage
- key is expected to be sent in the `Idempotency-Key` header (OpenAPI spec has been updated accordingly)

Resolves #185

### CODE REVIEW

File: openapi.yml [hunk 1]
```diff
@@ -112,7 +116,13 @@ paths:
               x-examples:
                 example-1:
                   maintenanceMode: false
-        description: 'Post only fields you want to be changed.'
+            examples:
```
Comment (nanuuki) [inline openapi.yml:119]: Courtesy of Stoplight Studio 🤷

File: docker-compose.yml [hunk 2]
```diff
@@ -10,6 +10,17 @@ services:
...
       - POSTGRES_PASSWORD=wallet
 
+  redis:
+    image: redis:6.2-alpine
+    restart: unless-stopped
+    command: redis-server /usr/local/etc/redis/redis.conf
+    ports:
+      - "6379:6379"
+    volumes:
+      - ./redis-data:/data
```
Comment (latenssi) [inline docker-compose.yml:20]: This should be a named volume IMO

File: .gitignore [hunk 3]
```diff
@@ -139,3 +139,4 @@ tmp/
 main
 
 flow/flowdb
+redis-data/
```
Comment (latenssi) [inline .gitignore:142]: Use a named volume instead and remove this

File: handlers/idempotency.go [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+		key := r.Header.Get("Idempotency-Key")
+		if len(key) == 0 && r.Method == http.MethodPost {
+			http.Error(rw, "Idempotency-Key header not found", http.StatusBadRequest)
+			return
+		}
+
+		exists, err := store.Get(key)
+		if err != nil {
+			panic(err)
```
Comment (latenssi) [inline handlers/idempotency.go:190]: Are we sure we want to panic in a http handler?

File: handlers/idempotency.go [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+		// XXX: same key w/ different payload should return 422,
+		// but isn't necessarily required functionality => only the key is stored
+		if exists {
+			http.Error(rw, fmt.Sprintf("Idempotency-Key conflict, key: %s", key), http.StatusConflict)
+			return
+		} else {
+			err := store.Set(key, opts.Expiry)
+			if err != nil {
+				panic(err)
```
Comment (latenssi) [inline handlers/idempotency.go:201]: Same here about the panic

**Review by latenssi — CHANGES_REQUESTED** [review]
Looking good, nice job! 

Only thing requesting for attention is the panics in the http handler. I think we should avoid panicing once the service is running but if you think it is necessary in that particular circumstance then I am ok with it.

File: handlers/idempotency.go (hunk 4 above)
Comment (nanuuki) [reply handlers/idempotency.go]: good catch :D no we most certainly don't!

**Review by latenssi — APPROVED** [review]

**Pull request merged**